File: code/comfyui/worker/java/src/main/java/dev/learn/comfy/worker/ComfyInstance.java
package dev.learn.comfy.worker;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.WebSocket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

import dev.learn.comfy.worker.Jobs.Job;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;
import tools.jackson.databind.json.JsonMapper;
import tools.jackson.databind.node.MissingNode;
import tools.jackson.databind.node.ObjectNode;

/**
 * One ComfyUI server: in production, one process per GPU, started with --cuda-device and its own --port.
 * The calls are lesson 4's, split so that the worker can retry, catch up and interrupt.
 */
public final class ComfyInstance {
    static final ObjectMapper JSON = JsonMapper.builder().build();

    private final String name;
    private final URI baseUri;
    private final HttpClient http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(3)).build();

    public ComfyInstance(String name, URI baseUri) {
        this.name = name;
        // resolve("prompt") needs a trailing slash to stay under the base URI
        this.baseUri = baseUri.toString().endsWith("/") ? baseUri : URI.create(baseUri + "/");
    }

    public String name() {
        return name;
    }

    public record GpuHealth(String name, boolean healthy, String device, long vramFree, int queueLength, String error) {
        @Override
        public String toString() {
            return healthy
                    ? name + ": healthy, device " + device + ", " + vramFree / (1024 * 1024) + " MiB free, " + queueLength + " prompt(s) in its queue"
                    : name + ": unreachable (" + error + ")";
        }
    }

    /** Numbers: each prompt's position number, which also appears in its history entry (prompt[0]). */
    public record QueueState(List<String> running, List<String> pending, Map<String, Double> numbers) {
        boolean contains(String promptId) {
            return running.contains(promptId) || pending.contains(promptId);
        }
    }

    public record Submitted(int status, String body) {
    }

    /** GET /system_stats and GET /queue, with a short timeout: a hung server counts as down. */
    public GpuHealth checkHealth() throws InterruptedException {
        try {
            JsonNode stats = getJson("system_stats", Duration.ofSeconds(3));
            JsonNode device = stats.path("devices").path(0);
            QueueState queue = getQueue(Duration.ofSeconds(3));
            return new GpuHealth(name, true, device.path("name").asString(), device.path("vram_free").asLong(),
                    queue.running().size() + queue.pending().size(), null);
        } catch (IOException | RuntimeException e) {
            return new GpuHealth(name, false, "", 0, 0, e.getClass().getSimpleName());
        }
    }

    /** GET /queue: each entry is [number, prompt_id, prompt, extra_data, outputs_to_execute]. */
    public QueueState getQueue(Duration timeout) throws IOException, InterruptedException {
        JsonNode queue = getJson("queue", timeout);
        var running = new ArrayList<String>();
        var pending = new ArrayList<String>();
        var numbers = new HashMap<String, Double>();
        for (JsonNode entry : queue.path("queue_running")) {
            running.add(entry.path(1).asString());
            numbers.merge(entry.path(1).asString(), entry.path(0).asDouble(), Math::max);
        }
        for (JsonNode entry : queue.path("queue_pending")) {
            pending.add(entry.path(1).asString());
            numbers.merge(entry.path(1).asString(), entry.path(0).asDouble(), Math::max);
        }
        return new QueueState(running, pending, numbers);
    }

    /** GET /history/{prompt_id}: null while the prompt is unknown, queued or running. */
    public JsonNode getHistory(String promptId) throws IOException, InterruptedException {
        JsonNode entry = getJson("history/" + promptId, Duration.ofSeconds(30)).path(promptId);
        return entry.isMissingNode() ? null : entry;
    }

    /** POST /prompt with the job id as prompt_id. The status code decides between retry and dead letter. */
    public Submitted submit(Job job, String clientId) throws IOException, InterruptedException {
        ObjectNode request = JSON.createObjectNode();
        request.set("prompt", job.workflow().deepCopy());
        request.put("client_id", clientId);
        request.put("prompt_id", job.id());
        HttpResponse<String> response = post("prompt", request, Duration.ofSeconds(30));
        return new Submitted(response.statusCode(), response.body());
    }

    /**
     * POST /interrupt for this prompt only, POST /queue to remove it if it hadn't started, then wait until it has
     * left the queue: a retry with the same prompt id must not receive this run's last messages.
     */
    public void cancelPrompt(String promptId) throws IOException, InterruptedException {
        long end = System.nanoTime() + Duration.ofSeconds(10).toNanos();
        post("interrupt", JSON.createObjectNode().put("prompt_id", promptId), Duration.ofSeconds(10));
        ObjectNode delete = JSON.createObjectNode();
        delete.putArray("delete").add(promptId);
        post("queue", delete, Duration.ofSeconds(10));
        while (getQueue(Duration.ofSeconds(10)).contains(promptId)) {
            if (System.nanoTime() > end) {
                throw new IOException("the prompt is still in the queue after 10 s");
            }
            Thread.sleep(50);
        }
    }

    public byte[] view(JsonNode image) throws IOException, InterruptedException {
        String query = "filename=" + encode(image.path("filename").asString())
                + "&subfolder=" + encode(image.path("subfolder").asString(""))
                + "&type=" + encode(image.path("type").asString("output"));
        HttpResponse<byte[]> response = http.send(HttpRequest.newBuilder(baseUri.resolve("view?" + query)).timeout(Duration.ofSeconds(60)).build(),
                HttpResponse.BodyHandlers.ofByteArray());
        if (response.statusCode() != 200) {
            throw new IOException("GET /view: " + response.statusCode());
        }
        return response.body();
    }

    /** GET /ws?clientId=…: connect before posting, reconnect with the same id after a drop. */
    public Socket connect(String clientId) throws IOException, InterruptedException {
        var messages = new LinkedBlockingQueue<JsonNode>();
        URI uri = URI.create(baseUri.toString().replaceFirst("^http", "ws") + "ws?clientId=" + clientId);
        try {
            WebSocket socket = http.newWebSocketBuilder().connectTimeout(Duration.ofSeconds(5)).buildAsync(uri, new Collector(messages)).join();
            return new Socket(socket, messages);
        } catch (CompletionException e) {
            throw new IOException("WebSocket connection failed: " + e.getCause(), e.getCause());
        }
    }

    /** The messages of one connection, in order. A missing node marks the end: closed or broken. */
    public record Socket(WebSocket webSocket, BlockingQueue<JsonNode> messages) {
        static final JsonNode CLOSED = MissingNode.getInstance();

        /** The next text message, null if none arrived within the timeout; throws once the connection is gone. */
        JsonNode receive(Duration timeout) throws IOException, InterruptedException {
            JsonNode message = messages.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
            if (message != null && message.isMissingNode()) {
                messages.add(message); // stays closed
                throw new IOException("the WebSocket is closed");
            }
            return message;
        }

        void close() {
            webSocket.abort();
        }
    }

    JsonNode getJson(String path, Duration timeout) throws IOException, InterruptedException {
        HttpResponse<String> response = http.send(HttpRequest.newBuilder(baseUri.resolve(path)).timeout(timeout).build(), HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new IOException("GET /" + path + ": " + response.statusCode());
        }
        return JSON.readTree(response.body());
    }

    HttpResponse<String> post(String path, JsonNode body, Duration timeout) throws IOException, InterruptedException {
        return http.send(HttpRequest.newBuilder(baseUri.resolve(path)).timeout(timeout)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(JSON.writeValueAsString(body))).build(), HttpResponse.BodyHandlers.ofString());
    }

    static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    /** Puts each complete text message in a queue; binary messages (previews) are skipped. */
    static final class Collector implements WebSocket.Listener {
        private final BlockingQueue<JsonNode> messages;
        private final StringBuilder text = new StringBuilder();

        Collector(BlockingQueue<JsonNode> messages) {
            this.messages = messages;
        }

        @Override
        public CompletionStage<?> onText(WebSocket socket, CharSequence data, boolean last) {
            text.append(data);
            if (last) {
                messages.add(JSON.readTree(text.toString()));
                text.setLength(0);
            }
            socket.request(1);
            return null;
        }

        @Override
        public CompletionStage<?> onBinary(WebSocket socket, ByteBuffer data, boolean last) {
            socket.request(1);
            return null;
        }

        @Override
        public CompletionStage<?> onClose(WebSocket socket, int statusCode, String reason) {
            messages.add(Socket.CLOSED);
            return null;
        }

        @Override
        public void onError(WebSocket socket, Throwable error) {
            messages.add(Socket.CLOSED);
        }
    }
}

File: code/comfyui/worker/csharp/ComfyWorker/ComfyInstance.cs
using System.Net;
using System.Net.Http.Json;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json.Nodes;

namespace Learn.Comfy.Worker;

/// <summary>
/// One ComfyUI server: in production, one process per GPU, started with --cuda-device and its own --port.
/// The calls are lesson 4's, split so that the worker can retry, catch up and interrupt.
/// </summary>
public sealed class ComfyInstance : IDisposable
{
    readonly HttpClient http;

    public ComfyInstance(string name, Uri baseUri, TimeSpan? requestTimeout = null)
    {
        Name = name;
        BaseUri = baseUri;
        http = new HttpClient { BaseAddress = baseUri, Timeout = requestTimeout ?? TimeSpan.FromSeconds(30) };
    }

    public string Name { get; }
    public Uri BaseUri { get; }

    /// <summary>GET /system_stats and GET /queue: is the server up, which device, how many prompts ahead.</summary>
    public async Task<GpuHealth> CheckHealthAsync(CancellationToken cancel)
    {
        // A probe answers quickly or counts as down: a hung server must not hold the scheduler for 30 seconds.
        using var probe = CancellationTokenSource.CreateLinkedTokenSource(cancel);
        probe.CancelAfter(TimeSpan.FromSeconds(3));
        try
        {
            var stats = await http.GetFromJsonAsync<JsonObject>("system_stats", probe.Token);
            var device = stats!["devices"]!.AsArray().FirstOrDefault();
            var queue = await GetQueueAsync(probe.Token);
            return new GpuHealth(Name, true, $"{device?["name"]}", device?["vram_free"]?.GetValue<long>() ?? 0, queue.Running.Count + queue.Pending.Count, null);
        }
        catch (Exception e) when ((e is HttpRequestException or TaskCanceledException or System.Text.Json.JsonException) && !cancel.IsCancellationRequested)
        {
            return new GpuHealth(Name, false, "", 0, 0, e.GetType().Name);
        }
    }

    /// <summary>GET /queue: each entry is [number, prompt_id, prompt, extra_data, outputs_to_execute].</summary>
    public async Task<QueueState> GetQueueAsync(CancellationToken cancel)
    {
        var queue = await http.GetFromJsonAsync<JsonObject>("queue", cancel);
        static List<string> Ids(JsonNode? entries) => entries!.AsArray().Select(e => e![1]!.GetValue<string>()).ToList();
        var numbers = queue!["queue_running"]!.AsArray().Concat(queue["queue_pending"]!.AsArray())
            .GroupBy(e => e![1]!.GetValue<string>())
            .ToDictionary(g => g.Key, g => g.Max(e => e![0]!.GetValue<double>()));
        return new QueueState(Ids(queue["queue_running"]), Ids(queue["queue_pending"]), numbers);
    }

    /// <summary>GET /history/{prompt_id}: {} while the prompt is unknown, queued or running.</summary>
    public async Task<JsonObject?> GetHistoryAsync(string promptId, CancellationToken cancel)
    {
        var history = await http.GetFromJsonAsync<JsonObject>($"history/{promptId}", cancel);
        return history?[promptId] as JsonObject;
    }

    /// <summary>POST /prompt with the job id as prompt_id. The status code decides between retry and dead letter.</summary>
    public async Task<(HttpStatusCode Status, string Body)> SubmitAsync(Job job, string clientId, CancellationToken cancel)
    {
        var request = new JsonObject { ["prompt"] = job.Workflow.DeepClone(), ["client_id"] = clientId, ["prompt_id"] = job.Id };
        using var response = await http.PostAsJsonAsync("prompt", request, cancel);
        return (response.StatusCode, await response.Content.ReadAsStringAsync(cancel));
    }

    /// <summary>POST /interrupt for this prompt only, POST /queue to remove it if it hadn't started, then wait
    /// until it has left the queue: a retry with the same prompt id must not receive this run's last messages.</summary>
    public async Task CancelPromptAsync(string promptId)
    {
        // Not the caller's token: this runs precisely when that token has fired.
        using var timeout = new CancellationTokenSource(TimeSpan.FromSeconds(10));
        using (await http.PostAsJsonAsync("interrupt", new JsonObject { ["prompt_id"] = promptId }, timeout.Token)) { }
        using (await http.PostAsJsonAsync("queue", new JsonObject { ["delete"] = new JsonArray(promptId) }, timeout.Token)) { }
        while ((await GetQueueAsync(timeout.Token)).Contains(promptId))
            await Task.Delay(50, timeout.Token);
    }

    public Task<byte[]> ViewAsync(JsonNode image, CancellationToken cancel)
    {
        string query = $"filename={Uri.EscapeDataString(image["filename"]!.GetValue<string>())}"
            + $"&subfolder={Uri.EscapeDataString(image["subfolder"]?.GetValue<string>() ?? "")}"
            + $"&type={Uri.EscapeDataString(image["type"]?.GetValue<string>() ?? "output")}";
        return http.GetByteArrayAsync($"view?{query}", cancel);
    }

    /// <summary>GET /ws?clientId=…: connect before posting, reconnect with the same id after a drop.</summary>
    public async Task<ClientWebSocket> ConnectAsync(string clientId, CancellationToken cancel)
    {
        var socket = new ClientWebSocket();
        var uri = new UriBuilder(BaseUri) { Scheme = BaseUri.Scheme == "https" ? "wss" : "ws", Path = "/ws", Query = $"clientId={clientId}" }.Uri;
        try
        {
            await socket.ConnectAsync(uri, cancel);
            return socket;
        }
        catch
        {
            socket.Dispose();
            throw;
        }
    }

    /// <summary>The next text message as JSON; binary messages (previews) are skipped.</summary>
    public static async Task<JsonNode> ReceiveJsonAsync(ClientWebSocket socket, CancellationToken cancel)
    {
        var buffer = new byte[16 * 1024];
        while (true)
        {
            using var message = new MemoryStream();
            WebSocketReceiveResult result;
            do
            {
                result = await socket.ReceiveAsync(buffer, cancel);
                if (result.MessageType == WebSocketMessageType.Close)
                    throw new WebSocketException(WebSocketError.ConnectionClosedPrematurely, "the server closed the WebSocket");
                message.Write(buffer, 0, result.Count);
            } while (!result.EndOfMessage);
            if (result.MessageType == WebSocketMessageType.Text)
                return JsonNode.Parse(Encoding.UTF8.GetString(message.ToArray()))!;
        }
    }

    public void Dispose() => http.Dispose();
}

/// <summary>Numbers: each prompt's position number, which also appears in its history entry (prompt[0]).</summary>
public sealed record QueueState(IReadOnlyList<string> Running, IReadOnlyList<string> Pending, IReadOnlyDictionary<string, double> Numbers)
{
    public bool Contains(string promptId) => Running.Contains(promptId) || Pending.Contains(promptId);
}

public sealed record GpuHealth(string Name, bool Healthy, string Device, long VramFree, int QueueLength, string? Error)
{
    public override string ToString() => Healthy
        ? $"{Name}: healthy, device {Device}, {VramFree / (1024 * 1024)} MiB free, {QueueLength} prompt(s) in its queue"
        : $"{Name}: unreachable ({Error})";
}
